### PR TITLE
Enhance CA handling 

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
## Objective
This PR enhances the system's ability to handle multiple Certificate Authorities (CAs) by modifying the logic to read multiple CA certificates from the configuration and concatenate them into a single file.

## Description
We implemented the same logic that allows the system to read multiple CA entries from the configuration and concatenate them into a single CA bundle file. This streamlined approach simplifies the setup process for scenarios requiring multiple CAs, this approach has been validated in https://github.com/wazuh/wazuh/pull/24620 through testing and revision

### CODE REVIEW

File: src/engine/source/indexerconnector/test/component/indexerConnector_test.cpp [hunk 1]
```diff
@@ -200,6 +203,28 @@ TEST_F(IndexerConnectorTest, ConnectionWithSslCredentials)
...
+    IndexerConnectorOptions indexerConfig {
+        .name = INDEXER_NAME,
+        .hosts = {A_ADDRESS},
+        .sslOptions = {.cacert = {"/etc/filebeat/certs/root-ca.pem", "/etc/filebeat/certs/root-ca-two.pem"},
+                       .cert = "/etc/filebeat/certs/filebeat.pem",
+                       .key = "/etc/filebeat/certs/filebeat-key.pem"},
+        .timeout = INDEXER_TIMEOUT};
+
+    // Create connector and wait until the connection is established.
+    // Throw is expected if the certs are not found.
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/test/component/indexerConnector_test.cpp:224]: I think a better option is to generate some dummy pem files and see if they were merged into one

**Review by sebasfalcone — CHANGES_REQUESTED** [review]

File: src/engine/source/indexerconnector/test/component/indexerConnector_test.cpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/test/component/indexerConnector_test.cpp:224]: I couldn't make it due to this check

https://github.com/wazuh/wazuh/blob/c1befeb4fc1399a8bd9552ea2a1daad83f18f012/src/engine/source/indexerconnector/src/indexerConnector.cpp#L82-L88

We could remove it or add the user and group on the test step, what do you think?

**Comment by GabrielEValenzuela** [comment]
# Testing locally 🟢 

```console
❯ src/engine/build/source/indexerconnector/test/component/indexerconnector_ctest --gtest_filter="*.ConnectionWithCertsArray"
Running main() from /home/.vcpkg/buildtrees/gtest/src/v1.15.0-3c2b6dd10b.clean/googletest/src/gtest_main.cc
Note: Google Test filter = *.ConnectionWithCertsArray
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from IndexerConnectorTest
[ RUN      ] IndexerConnectorTest.ConnectionWithCertsArray
getpwnam: wazuh //Used only for debug
getgrnam: wazuh //Used only for debug
2024-10-18 16:54:30.518 888274:888274 warning: No username found in the configuration, using default value.
2024-10-18 16:54:30.518 888274:888274 warning: No password found in the configuration, using default value.
[       OK ] IndexerConnectorTest.ConnectionWithCertsArray (22 ms)
[----------] 1 test from IndexerConnectorTest (22 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (22 ms total)
[  PASSED  ] 1 test.
❯ ctest --test-dir src/engine/build/source/indexerconnector/ -T test -C release --output-on-failure
Internal ctest changing into directory: /home/gvalenzuela/Documents/Work/wazuh/src/engine/build/source/indexerconnector
Cannot find file: /home/gvalenzuela/Documents/Work/wazuh/src/engine/build/source/indexerconnector/DartConfiguration.tcl
   Site: 
   Build name: (empty)
Cannot find file: /home/gvalenzuela/Documents/Work/wazuh/src/engine/build/source/indexerconnector/DartConfiguration.tcl
Test project /home/gvalenzuela/Documents/Work/wazuh/src/engine/build/source/indexerconnector
      Start  1: MonitoringTest.TestCheckIfAnUnregisteredServerIsAvailable
 1/25 Test  #1: MonitoringTest.TestCheckIfAnUnregisteredServerIsAvailable .......   Passed    0.02 sec
      Start  2: MonitoringTest.TestInstantiationWithGreenRedServers
 2/25 Test  #2: MonitoringTest.TestInstantiationWithGreenRedServers .............   Passed    0.01 sec
      Start  3: MonitoringTest.TestInstantiationWithoutServers
 3/25 Test  #3: MonitoringTest.TestInstantiationWithoutServers ..................   Passed    0.01 sec
      Start  4: ServerSelectorTest.TestInstantiation
 4/25 Test  #4: ServerSelectorTest.TestInstantiation ............................   Passed    0.01 sec
      Start  5: ServerSelectorTest.TestInstantiationWithoutServers
 5/25 Test  #5: ServerSelectorTest.TestInstantiationWithoutServers ..............   Passed    0.01 sec
      Start  6: ServerSelectorTest.TestGetNextBeforeHealthCheck
 6/25 Test  #6: ServerSelectorTest.TestGetNextBeforeHealthCheck .................   Passed    0.01 sec
      Start  7: ServerSelectorTest.TestGetNextBeforeAndAfterHealthCheck
 7/25 Test  #7: ServerSelectorTest.TestGetNextBeforeAndAfterHealthCheck .........   Passed    0.01 sec
      Start  8: ServerSelectorTest.TestGextNextWhenThereAreNoAvailableServers
 8/25 Test  #8: ServerSelectorTest.TestGextNextWhenThereAreNoAvailableServers ...   Passed    0.01 sec
      Start  9: IndexerConnectorTest.ConnectionWithUserAndPassword
 9/25 Test  #9: IndexerConnectorTest.ConnectionWithUserAndPassword ..............   Passed    0.05 sec
      Start 10: IndexerConnectorTest.ConnectionWithSslCredentials
10/25 Test #10: IndexerConnectorTest.ConnectionWithSslCredentials ...............   Passed    0.03 sec
      Start 11: IndexerConnectorTest.ConnectionWithCertsArray
11/25 Test #11: IndexerConnectorTest.ConnectionWithCertsArray ...................   Passed    0.02 sec
      Start 12: IndexerConnectorTest.ConnectionWithCertsArrayNoFiles
12/25 Test #12: IndexerConnectorTest.ConnectionWithCertsArrayNoFiles ............   Passed    0.01 sec
      Start 13: IndexerConnectorTest.ConnectionUnavailableServer
13/25 Test #13: IndexerConnectorTest.ConnectionUnavailableServer ................   Passed    2.13 sec
      Start 14: IndexerConnectorTest.ConnectionMultipleServers
14/25 Test #14: IndexerConnectorTest.ConnectionMultipleServers ..................   Passed    0.03 sec
      Start 15: IndexerConnectorTest.ConnectionInvalidServer
15/25 Test #15: IndexerConnectorTest.ConnectionInvalidServer ....................   Passed    2.13 sec
      Start 16: IndexerConnectorTest.Publish
16/25 Test #16: IndexerConnectorTest.Publish ....................................   Passed    5.07 sec
      Start 17: IndexerConnectorTest.PublishDeleted
17/25 Test #17: IndexerConnectorTest.PublishDeleted .............................   Passed    5.05 sec
      Start 18: IndexerConnectorTest.PublishWithoutId
18/25 Test #18: IndexerConnectorTest.PublishWithoutId ...........................   Passed    5.04 sec
      Start 19: IndexerConnectorTest.PublishUnavailableServer
19/25 Test #19: IndexerConnectorTest.PublishUnavailableServer ...................   Passed    7.65 sec
      Start 20: IndexerConnectorTest.PublishInvalidData
20/25 Test #20: IndexerConnectorTest.PublishInvalidData .........................   Passed    7.64 sec
      Start 21: IndexerConnectorTest.DiscardInvalidJSON
21/25 Test #21: IndexerConnectorTest.DiscardInvalidJSON .........................   Passed    7.64 sec
      Start 22: IndexerConnectorTest.PublishTwoIndexes
22/25 Test #22: IndexerConnectorTest.PublishTwoIndexes ..........................   Passed   10.04 sec
      Start 23: IndexerConnectorTest.PublishErrorFromServer
23/25 Test #23: IndexerConnectorTest.PublishErrorFromServer .....................   Passed    5.04 sec
      Start 24: IndexerConnectorTest.UpperCaseCharactersIndexName
24/25 Test #24: IndexerConnectorTest.UpperCaseCharactersIndexName ...............   Passed    0.01 sec
      Start 25: IndexerConnectorTest.PublishDatePlaceholder
25/25 Test #25: IndexerConnectorTest.PublishDatePlaceholder .....................   Passed    5.03 sec

100% tests passed, 0 tests failed out of 25

Total Test time (real) =  62.70 sec
```

File: src/engine/source/indexerconnector/test/component/indexerConnector_test.cpp (hunk 1 above)
Comment (sebasfalcone) [reply src/engine/source/indexerconnector/test/component/indexerConnector_test.cpp:321]: Yes I think we are okay with that

**Review by sebasfalcone — APPROVED** [review]
LGTM

- I've liked the mocking strategy 👍🏽 


## Testing
I've runned the test tool with the following configuration
```json
{
  "name": "wazuh-states-vulnerabilities-cluster",
  "enabled": "yes",
  "hosts": ["http://0.0.0.0:9200"],
  "ssl": {
    "certificate_authorities": [
      "/workspaces/dev-tools/dev-containers/wazuh/wazuh/src/engine/build/source/indexerconnector/tool/ca1",
      "/workspaces/dev-tools/dev-containers/wazuh/wazuh/src/engine/build/source/indexerconnector/tool/ca2"
    ]
  }
}
```

- Logs
```
2024-10-18 22:13:22.560 8581:8581 indexerConnector.cpp:95 at mergeCaRootCertificates(): debug: All CA files merged into 'tmp/root-ca-merged.pem' successfully.
```

- Indiviual certs:
```
cat ca1
1testing1

cat ca2
2testing2
```

- Merged cert:
```
cat tmp/root-ca-merged.pem 
1testing1
2testing2
```

>[!NOTE]
> I've added the wazuh user and group in order to run the test
>
> ```
> sudo useradd wazuh     
> sudo groupadd wazuh      
> sudo usermod -aG wazuh wazuh  
> ```

File: src/engine/source/indexerconnector/src/indexerConnector.cpp [hunk 2]
```diff
@@ -9,23 +9,92 @@
...
+
 #include "secureCommunication.hpp"
 #include "serverSelector.hpp"
-#include <fstream>
 
 constexpr auto INDEXER_COLUMN {"indexer"};
 constexpr auto USER_KEY {"username"};
 constexpr auto PASSWORD_KEY {"password"};
 constexpr auto ELEMENTS_PER_BULK {1000};
+constexpr auto WAZUH_OWNER {"wazuh"};
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/src/indexerConnector.cpp:31]:
```suggestion
constexpr auto WAZUH_OWNER {"wazuh"};
constexpr auto WAZUH_GROUP {"wazuh"};
```

File: src/engine/source/indexerconnector/src/indexerConnector.cpp [hunk 3]
```diff
@@ -9,23 +9,92 @@
...
+        std::ifstream file(filePath);
+        if (!file.is_open())
+        {
+            throw std::runtime_error("Could not open CA root certificate file: '" + filePath + "'.");
+        }
+
+        caRootCertificateContentMerged.append((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+    }
+
+    caRootCertificate = DEFAULT_PATH;
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/src/indexerConnector.cpp:65]: This cannot be configured so they are the only path

File: src/engine/source/indexerconnector/src/indexerConnector.cpp [hunk 4]
```diff
@@ -9,23 +9,92 @@
...
 #include "secureCommunication.hpp"
 #include "serverSelector.hpp"
-#include <fstream>
 
 constexpr auto INDEXER_COLUMN {"indexer"};
 constexpr auto USER_KEY {"username"};
 constexpr auto PASSWORD_KEY {"password"};
 constexpr auto ELEMENTS_PER_BULK {1000};
+constexpr auto WAZUH_OWNER {"wazuh"};
+constexpr auto DEFAULT_PATH {"tmp/root-ca-merged.pem"};
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/src/indexerConnector.cpp:32]:
```suggestion
constexpr auto MERGED_CA_PATH {"tmp/root-ca-merged.pem"};
```

**Review by sebasfalcone — CHANGES_REQUESTED** [review]

File: src/engine/source/indexerconnector/src/indexerConnector.cpp (hunk 2 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/src/indexerConnector.cpp:31]: Done, thanks!

File: src/engine/source/indexerconnector/src/indexerConnector.cpp (hunk 4 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/src/indexerConnector.cpp]: Done, thanks!

File: src/engine/source/indexerconnector/src/indexerConnector.cpp (hunk 3 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/src/indexerConnector.cpp]: Done, thanks!

File: src/engine/source/indexerconnector/src/indexerConnector.cpp [hunk 5]
```diff
@@ -9,23 +9,93 @@
...
 // Single thread in case the events needs to be processed in order.
 constexpr auto SINGLE_ORDERED_DISPATCHING = 1;
 
+/**
+ * @brief Merges the CA root certificates into a single file.
+ * @param filePaths The list of CA root certificates file paths.
+ * @param caRootCertificate The path to the merged CA root certificate.
+ * @throws std::runtime_error If the CA root certificate file does not exist, could not be opened, written or the
+ * ownership could not be changed.
+ * @note The merged file will be created in the `tmp` directory of Wazuh installation.
```
Comment (sebasfalcone) [inline src/engine/source/indexerconnector/src/indexerConnector.cpp:44]:
```suggestion
```

**Review by sebasfalcone — APPROVED** [review]

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/engine/source/indexerconnector/src/indexerConnector.cpp [hunk 6]
```diff
@@ -9,23 +9,92 @@
...
 #include "serverSelector.hpp"
-#include <fstream>
 
 constexpr auto INDEXER_COLUMN {"indexer"};
 constexpr auto USER_KEY {"username"};
 constexpr auto PASSWORD_KEY {"password"};
 constexpr auto ELEMENTS_PER_BULK {1000};
+constexpr auto WAZUH_OWNER {"wazuh"};
+constexpr auto WAZUH_GROUP {"wazuh"};
+constexpr auto MERGED_CA_PATH {"tmp/wazuh-server/root-ca-merged.pem"};
```
Comment (Dwordcito) [inline src/engine/source/indexerconnector/src/indexerConnector.cpp:33]:
```suggestion
constexpr auto MERGED_CA_PATH {"/tmp/wazuh-server/root-ca-merged.pem"};
```

**Review by Dwordcito — CHANGES_REQUESTED** [review]

File: src/engine/source/indexerconnector/src/indexerConnector.cpp (hunk 6 above)
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/src/indexerConnector.cpp]: Solved!

File: src/engine/source/indexerconnector/src/indexerConnector.cpp [hunk 7]
```diff
@@ -34,7 +103,11 @@ static void initConfiguration(SecureCommunication& secureCommunication, const In
...
 
-    if (!config.sslOptions.cacert.empty())
+    if (config.sslOptions.cacert.size() > 1)
+    {
+        mergeCaRootCertificates(config.sslOptions.cacert, caRootCertificate);
+    }
+    else if (!config.sslOptions.cacert.empty())
     {
         caRootCertificate = config.sslOptions.cacert.front();
     }
```
[the remark this answers is not in the document]
Comment (GabrielEValenzuela) [reply src/engine/source/indexerconnector/src/indexerConnector.cpp:113]: Solved!

**Review by Dwordcito — APPROVED** [review]

**Pull request merged**